# EC2202 Python Primer (cont'd)

## Student Information

* Student ID: 20225186
* Name: 주서현

**Disclaimer.**
This Python primer is based on

1. [Stanford CS231n Python Tutorial](https://colab.research.google.com/github/cs231n/cs231n.github.io/blob/master/python-colab.ipynb)
2. [LearnPython.org](https://www.learnpython.org/en/Welcome)

## Introduction

In this Python primer (cont'd), we will cover:
* Containers
* Functions
* Map, filter and reduce
* Higher order functions
* Style guide for python code

In [ ]:
%%HTML
<iframe width="560" height="315" src="https://www.youtube.com/embed/RbBMn-868iQ" title="YouTube video player" frameborder="0" allowfullscreen></iframe>
<iframe width="560" height="315" src="https://www.youtube.com/embed/_0-2NMPUViM" title="YouTube video player" frameborder="0" allowfullscreen></iframe>

## Containers


Python includes several built-in container types: lists, dictionaries, sets, and tuples. We will start with list comprehensions.

### Lists: Skip this part

In [ ]:
# Slicing
odds = [3, 5, 7, 9, 11]
a_list = list(range(1, 5))
print(len([odds[i] for i in range(1, 5)]))
print(max(a_list))
print(min(a_list))

# Aggregation
print(sum(odds))
print(sum({3:9, 5:25}))
max(range(10))
max(range(10), key=lambda x: 7 - (x-2)*(x-4))
all([x < 5 for x in range(5)])
perfect_square = lambda x: x == round(x ** 0.5) ** 2
any([perfect_square(x) for x in range(50, 60)])

4
4
1
35
8


False

### **List Comprehensions**

When programming, frequently we want to transform one type of data into another. As a simple example, consider the following code that computes square numbers:

In [ ]:
nums = [0, 1, 2, 3, 4]
squares = []
for x in nums:
    squares.append(x ** 2)
print(squares)

[0, 1, 4, 9, 16]


In [2]:
nums = [0, 1, 2, 3, 4]
squares = []

for x in nums:
  squares.append(x**2)


print(squares)

# the above procedure can be simplified into the following expression

nums = [0, 1, 2, 3, 4]
squares = [x**2 for x in nums]

print(squares)

[0, 1, 4, 9, 16]
[0, 1, 4, 9, 16]


You can make this code simpler using a list comprehension:

In [ ]:
nums = [0, 1, 2, 3, 4]
squares = [x ** 2 for x in nums]
print(squares)

[0, 1, 4, 9, 16]


List comprehensions can also contain conditions:

In [ ]:
nums = [0, 1, 2, 3, 4]
even_squares = [x ** 2 for x in nums if x % 2 == 0]
print(even_squares)

Another example:

In [ ]:
sentence = "the quick brown fox jumps over the lazy dog"
words = sentence.split()
word_lengths = []
for word in words:
      if word != "the":
          word_lengths.append(len(word))
print(words)
print(word_lengths)

In [5]:
sentence = "this is only just the beginning"
words = sentence.split()
#print(words); ['this', 'is', 'only', 'just', 'the', 'beginning']
word_lengths = []

for word in words:
  if word != 'the':
    word_lengths.append(len(word))

print(words)
print(word_lengths)

['this', 'is', 'only', 'just', 'the', 'beginning']
[4, 2, 4, 4, 9]


In [6]:
sentence = "this is only just the beginning"
words = sentence.split()

word_lengths = [len(word) for word in words if word !='the']
print(word_lengths)

[4, 2, 4, 4, 9]


In [ ]:
sentence = "the quick brown fox jumps over the lazy dog"
words = sentence.split()
word_lengths = [len(word) for word in words if word != "the"]
print(words)
print(word_lengths)

You can also iterate over two iterables as follows

(두번 돌리기 가능) (중첩)

In [7]:
seq_x = [1, 2, 3, 4]
seq_y = 'abc'
print([(x,y) for x in seq_x for y in seq_y])

[(1, 'a'), (1, 'b'), (1, 'c'), (2, 'a'), (2, 'b'), (2, 'c'), (3, 'a'), (3, 'b'), (3, 'c'), (4, 'a'), (4, 'b'), (4, 'c')]


In [9]:
seq_x = list(range(1, 6))
seq_y = list(range(100,700, 100))
print([(x,y) for x in seq_x for y in seq_y])

print(seq_y)

[(1, 100), (1, 200), (1, 300), (1, 400), (1, 500), (1, 600), (2, 100), (2, 200), (2, 300), (2, 400), (2, 500), (2, 600), (3, 100), (3, 200), (3, 300), (3, 400), (3, 500), (3, 600), (4, 100), (4, 200), (4, 300), (4, 400), (4, 500), (4, 600), (5, 100), (5, 200), (5, 300), (5, 400), (5, 500), (5, 600)]
[100, 200, 300, 400, 500, 600]


In [16]:
seq_x = [1, 2, 3, 4] #list
seq_y = 'abc' #string

empty_seq = [(x,y) for x in seq_x for y in seq_y]
print(empty_seq)
print(y for y in seq_y)
print([y for y in seq_y])

[(1, 'a'), (1, 'b'), (1, 'c'), (2, 'a'), (2, 'b'), (2, 'c'), (3, 'a'), (3, 'b'), (3, 'c'), (4, 'a'), (4, 'b'), (4, 'c')]
<generator object <genexpr> at 0x7c483b8cf290>
['a', 'b', 'c']


It also supports both “if” statements and referencing the outer iterator from the inner one, like so:

In [ ]:
text_in_text = [["Hello", "World!"], ["Lets", "Eat!"]]
print([word for words in text_in_text for word in words])

**WWPP Question**

In [ ]:
seq = ['abc', 'def', 'g', 'hi']
print([y for x in seq if len(x) > 1 for y in x if y != 'e'])

"""
same as...

empty_seq = []
for x in seq:
  if len(x) >1:
    for y in x:
      if y != 'e':
        empty_seq.append(y)
"""

['a', 'b', 'c', 'd', 'f', 'h', 'i']


**Practice Questions**

Q1) Find all numbers from 1 to 100 that are divisible by 7

Q2) Find all numbers from 1 to 100 that have a 3 in them

Q3) Count the number of spaces in "the quick brown fox jumps over the lazy dog"

Q4) Create a list of all the consonants in the string "Yellow Yaks like yelling and yawning and yesturday they yodled while eating yuky yams"

Q5) Find the common items in two lists
- list_a = [1, 2, 3, 4, 5, 7, 10]
- list_b = [3, 7, 8, 9, 11, 12]



In [25]:
q1 = [n for n in range(1,101) if n % 7 == 0]
q2 = [n for n in range(1,101) if '3' in str(n)]
q3 = sum([1 for l in "the quick brown fox jumps over the lazy dog" if l == ' '])
sentence = 'yellow yaks like yelling and yawning and yesterday they yodled while eating yuky yams'
q4 = [l for l in sentence if l not in 'a, e, i, o, u']
list_a = [1, 2, 3, 4, 5, 7, 10]
list_b = [3, 7, 8, 9, 11, 12]
q5 = [a for a in list_a if a in list_b]

print(q1)
print(q2)
print(q3)
print(q4)
print(q5)

[7, 14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84, 91, 98]
[3, 13, 23, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 43, 53, 63, 73, 83, 93]
8
['y', 'l', 'l', 'w', 'y', 'k', 's', 'l', 'k', 'y', 'l', 'l', 'n', 'g', 'n', 'd', 'y', 'w', 'n', 'n', 'g', 'n', 'd', 'y', 's', 't', 'r', 'd', 'y', 't', 'h', 'y', 'y', 'd', 'l', 'd', 'w', 'h', 'l', 't', 'n', 'g', 'y', 'k', 'y', 'y', 'm', 's']
[3, 7]


In [ ]:
%%HTML
<iframe width="560" height="315" src="https://www.youtube.com/embed/F-oayUtAgfM" title="YouTube video player" frameborder="0" allowfullscreen></iframe>
<iframe width="560" height="315" src="https://www.youtube.com/embed/t2MchlFNptE" title="YouTube video player" frameborder="0" allowfullscreen></iframe>

### Dictionaries

A dictionary is a data type similar to arrays, but works with (key, value) pairs, similar to a `Map` in Java or an `object` in Javascript. Each value stored in a dictionary can be accessed using a key, which is any type of object (a string, a number, a list, etc.) instead of using its index to address it.

In [26]:
d = {'cat': 'cute', 'dog': 'furry'}  # Create a new dictionary with some data
print(d['cat'])                      # Get an entry from a dictionary; prints "cute"
print('cat' in d)                    # Check if a dictionary has a given key; prints "True"

cute
True


In [29]:
d = {'cat': 'cute', 'dog':'furry'}
print(d['cat'])
print('cat' in d)

cute
True


In [27]:
d['fish'] = 'wet'  # Set an entry in a dictionary
print(d['fish'])   # Prints "wet"

wet


In [32]:
d['fish'] = 'wet'
print(d['fish'])
print(d)
print(type(d))

wet
{'cat': 'cute', 'dog': 'furry', 'fish': 'wet'}
<class 'dict'>


In [33]:
print(d['monkey'])  # KeyError: 'monkey' not a key of d

KeyError: 'monkey'

In [ ]:
print(d.get('monkey', 'N/A'))  # Get an element with a default; prints "N/A"
print(d.get('fish', 'N/A'))    # Get an element with a default; prints "wet"
print(d)

In [36]:
print(d.get('monkey', 'N/A'))
print(d.get('fish', 'N/A'))
print(d.get('dune', 'awesome'))
print(d)

N/A
wet
awesome
{'cat': 'cute', 'dog': 'furry', 'fish': 'wet'}


In [ ]:
numerals = {'I': 1, 'V': 5, 'X': 10}
numerals['X']
# numerals['X-ray']
# numerals[10]
len(numerals)
list(numerals)
numerals.values()
list(numerals.values())
sum(numerals.values())
dict([[3, 9], [4, 16]])
numerals.get('X', 0)
numerals.get('X-ray', 0)
{1: 2, 1: 3}
{[1]: 2}
{1: [2]}

In [4]:
numerals = {'l':1, 'v':5, 'x':10}
print(numerals['x'])
print(len(numerals))
print(list(numerals)) #key만 리스트에 저장됨

10
3
['l', 'v', 'x']


In [6]:
print(numerals.values())

dict_values([1, 5, 10])


In [7]:
del numerals['x']
print(numerals)

{'l': 1, 'v': 5}


To remove a specified index, use either one of the following notations:

In [ ]:
del d['fish']        # Remove an element from a dictionary
# d.pop('fish')
print(d)
print(d.get('fish', 'N/A')) # "fish" is no longer a key; prints "N/A"
print(d)

You can find all you need to know about dictionaries in the [documentation](https://docs.python.org/2/library/stdtypes.html#dict).

It is easy to iterate over the keys in a dictionary:

In [10]:
d = {'person': 2, 'cat': 4, 'spider': 8}
for animal, legs in d.items():
    print('A {} has {} legs'.format(animal, legs))

A person has 2 legs
A cat has 4 legs
A spider has 8 legs


Dictionary comprehensions: These are similar to list comprehensions, but allow you to easily construct dictionaries. For example:

In [9]:
nums = [0, 1, 2, 3, 4]
even_num_to_square = {x: x ** 2 for x in nums if x % 2 == 0}
print(even_num_to_square)

{0: 0, 2: 4, 4: 16}


### Sets

A set is an unordered collection of distinct elements. As a simple example, consider the following:

In [11]:
animals = {'cat', 'dog'}
print('cat' in animals)   # Check if an element is in a set; prints "True"
print('fish' in animals)  # prints "False"

True
False


In [12]:
animals.add('fish')      # Add an element to a set
print('fish' in animals)
print(len(animals))       # Number of elements in a set;

True
3


In [13]:
animals.add('cat')       # Adding an element that is already in the set does nothing
print(len(animals))
animals.remove('cat')    # Remove an element from a set
print(len(animals))

3
2


Loops: Iterating over a set has the same syntax as iterating over a list; however since sets are unordered, you cannot make assumptions about the order in which you visit the elements of the set:

In [14]:
animals = {'cat', 'dog', 'fish'}
for idx, animal in enumerate(animals):
    print('#{}: {}'.format(idx + 1, animal))

#1: fish
#2: cat
#3: dog


In [15]:
animals = {'cat', 'dog', 'fish'}

for n, animal in enumerate(animals):
  print('#{}: {}'.format(n+1, animal))

#1: fish
#2: cat
#3: dog


set에서는 (일단은) 출력될때 순서가 이상(?)함.

Set comprehensions: Like lists and dictionaries, we can easily construct sets using set comprehensions:

In [ ]:
from math import sqrt
print({int(sqrt(x)) for x in range(30)})

In [17]:
from math import sqrt
print({int(sqrt(x)) for x in range(30)})

{0, 1, 2, 3, 4, 5}


### Tuples

A tuple is an (***immutable***) ordered list of values. A tuple is in many ways similar to a list; one of the most important differences is that tuples can be used as keys in dictionaries and as elements of sets, while lists cannot. Here is a trivial example:

In [ ]:
d = {(x, x + 1): x for x in range(10)}  # Create a dictionary with tuple keys
t = (5, 6)       # Create a tuple
print(type(t))
print(d[t])
print(d[(1, 2)])

In [20]:
d = {(x, x+1): x for x in range(10)}

t = (5, 6, 7)
print(type(t))
print(type(d))
print(d)
print(d[(1, 2)])

<class 'tuple'>
<class 'dict'>
{(0, 1): 0, (1, 2): 1, (2, 3): 2, (3, 4): 3, (4, 5): 4, (5, 6): 5, (6, 7): 6, (7, 8): 7, (8, 9): 8, (9, 10): 9}
1


## Functions


Python functions are defined using the `def` keyword. For example:

In [ ]:
def sign(x):
    if x > 0:
        return 'positive'
    elif x < 0:
        return 'negative'
    else:
        return 'zero'

for x in [-1, 0, 1]:
    print(sign(x))

In [25]:
def sign(x):
  if x>0: return 'positive'
  elif x<0: return 'negative'
  else: return 'zero'

for x in range(-5, 6): print('{}: {}'.format(x, sign(x)))

-5: negative
-4: negative
-3: negative
-2: negative
-1: negative
0: zero
1: positive
2: positive
3: positive
4: positive
5: positive


We will often define functions to take optional keyword arguments, like this:

In [27]:
def hello(name, loud=False):
    if loud:
        print('HELLO, {}'.format(name.upper()))
    else:
        print('Hello, {}!'.format(name))

hello('Bob')
hello('Fred', loud=True)

Hello, Bob!
HELLO, FRED


이 함수는 hello라는 이름을 가진 함수로, 인자로 name과 loud를 받습니다. loud의 기본값은 False로 설정되어 있으므로, 인자를 전달하지 않으면 loud는 False로 간주됩니다. 이 함수는 loud가 True로 설정되면 이름을 대문자로 출력하고, 그렇지 않으면 소문자로 출력합니다.

In [31]:
def hello(name, loud=False):
  if loud:
    print('HELLO, {}'.format(name.upper()))
  else:
    print('Hello, {}!'.format(name))

hello('bob', loud=False)
hello('amy', loud=True)

Hello, bob!
HELLO, AMY


In [37]:
def num_digits(n):
  length_n = 0
  while n>0:
    n = n//10
    length_n +=1
  return length_n

def same_length(a, b):
  return num_digits(a) == num_digits(b)

print(same_length(321, 12345))

False


**WWPP Question**

In [40]:
def bake(cake, make):
  if cake == 0:
    cake = cake + 1
    print(cake)
  if cake == 1:
    print(make)
  else:
    return cake
  return make

print(bake(0, 29)) #(cake = 0, print cake(0->1), cake = 1, print(make), return make)
print('--------')
print(bake(1, "mashed potatoes"))

1
29
29
--------
mashed potatoes
mashed potatoes


In [ ]:
%%HTML
<iframe width="560" height="315" src="https://www.youtube.com/embed/mt9Ox2bdv58" title="YouTube video player" frameborder="0" allowfullscreen></iframe>
<iframe width="560" height="315" src="https://www.youtube.com/embed/Xn_sESmq1T4" title="YouTube video player" frameborder="0" allowfullscreen></iframe>

We generally let a function do one task. This is sometimes called *modular design* or DRY (don't repeat yourself!).

In [ ]:
def same_length(a, b):
    """Return whether positive integers a and b have the same number of digits.

    >>> same_length(50, 70)
    True
    >>> same_length(50, 100)
    False
    >>> same_length(1000, 100000)
    False
    """
    return num_digits(a) == num_digits(b)
    # a_digits = 0
    # while a > 0:
    #     a = a // 10
    #     a_digits = a_digits + 1
    # b_digits = 0
    # while b > 0:
    #     b = b // 10
    #     b_digits = b_digits + 1
    # return a_digits == b_digits


def num_digits(a):
    a_digits = 0
    while a > 0:
        a = a // 10
        a_digits = a_digits + 1
    return a_digits

## Map, Filter and Reduce

Map, Filter, and Reduce are paradigms of functional programming. They allow the programmer (you) to write simpler, shorter code, without neccessarily needing to bother about intricacies like loops and branching.

Essentially, these three functions allow you to apply a function across a number of iterables, in one fell swoop. `map` and `filter` come built-in with Python (in the \_\_builtins\_\_ module) and require no importing. `reduce`, however, needs to be imported as it resides in the `functools` module. Let's get a better understanding of how they all work, starting with `map`.

### `map`

The `map()` function in python has the following syntax:

`map(func, *iterables)`

Where `func` is the function on which each element in `iterables` (as many as they are) would be applied on. Notice the asterisk(`*`) on `iterables`? It means there can be as many iterables as possible, in so far `func` has that exact number as required input arguments. Before we move on to an example, it's important that you note the following:

1. In Python 2, the `map()` function returns a list. In Python 3, however, the function returns a `map object` which is a generator object. To get the result as a list, the built-in `list()` function can be called on the map object. i.e. list(map(func, *iterables))
2. The number of arguments to `func` must be the number of `iterables` listed.
Let's see how these rules play out with the following examples.

Say I have a list (`iterable`) of my favourite pet names, all in lower case and I need them in uppercase. Traditonally, in normal pythoning, I would do something like this:

In [43]:
my_pets = ['alfred', 'tabitha', 'william', 'arla']
uppered_pets = []

for pet in my_pets:
    pet_ = pet.upper()
    uppered_pets.append(pet_)

print(uppered_pets)

['ALFRED', 'TABITHA', 'WILLIAM', 'ARLA']


With `map()` functions, it's not only easier, but it's also much more flexible. I simply do this:

In [41]:
def str_uppper(string):
  return string.upper()

my_pets = ['alfred', 'tabitha', 'william', 'arla']
uppered_pets = list(map(str_uppper, my_pets))

print(uppered_pets)

['ALFRED', 'TABITHA', 'WILLIAM', 'ARLA']


In [50]:
my_pets = ['alfred', 'tabitha', 'william', 'arla']
upp_pets = list(map(str.upper, my_pets))

print(uppered_pets)

['ALFRED', 'TABITHA', 'WILLIAM', 'ARLA']


In [42]:
my_pets = ['alfred', 'tabitha', 'william', 'arla']
uppered_pets = list(map(str.upper, my_pets))

print(uppered_pets)

['ALFRED', 'TABITHA', 'WILLIAM', 'ARLA']


In [56]:
def square(x):
  return x**2

numbers = [1, 2, 3, 4, 5]
sq_nums = map(square, numbers) # map(function, iterable)

print(list(sq_nums))

[1, 4, 9, 16, 25]


Note that using the defined `map()` syntax above, `func` in this case is `str.upper` and `iterables` is the `my_pets` list -- just one iterable. Also note that we did not call the `str.upper` function (doing this: `str.upper()`), as the map function does that for us on each element in the `my_pets` list.

What's more important to note is that the `str.upper` function requires only one argument by definition and so we passed just one iterable to it. So, if the function you're passing requires two, or three, or n arguments, then you need to pass in two, three or n iterables to it. Let me clarify this with another example.

Say I have a list of circle areas that I calculated somewhere, all in five decimal places. And I need to round each element in the list up to its position decimal places, meaning that I have to round up the first element in the list to one decimal place, the second element in the list to two decimal places, the third element in the list to three decimal places, etc. With `map()` this is a piece of cake. Let's see how.

Python already blesses us with the `round()` built-in function that takes two arguments -- the number to round up and the number of decimal places to round the number up to. So, since the function requires two arguments, we need to pass in **two** iterables.

In [57]:
print(round(3.1415923123, 1))
print(round(3.1415923123, 3))
print(round(3.1415923123, 5))

circle_areas = [3.56773, 5.57668, 4.00914, 56.24241, 9.01344, 32.00013]
result = list(map(round, circle_areas, range(1, 7)))
print(result)

result = list(map(round, circle_areas, [5] * 6))
print(result)

3.1
3.142
3.14159
[3.6, 5.58, 4.009, 56.2424, 9.01344, 32.00013]
[3.56773, 5.57668, 4.00914, 56.24241, 9.01344, 32.00013]


See the beauty of `map()`? Can you imagine the flexibility this evokes?

The `range(1, 7)` function acts as the second argument to the `round` function (the number of required decimal places per iteration). So as `map` iterates through `circle_areas`, during the first iteration, the first element of `circle_areas`, `3.56773` is passed along with the first element of `range(1,7)`, `1` to `round`, making it effectively become `round(3.56773, 1)`. During the second iteration, the second element of `circle_areas`, `5.57668` along with the second element of `range(1,7)`, `2` is passed to `round` making it translate to `round(5.57668, 2)`. This happens until the end of the `circle_areas` list is reached.

I'm sure you're wondering: "What if I pass in an iterable less than or more than the length of the first iterable? That is, what if I pass `range(1, 3)` or `range(1, 9999)` as the second iterable in the above function". And the answer is simple: nothing! Okay, that's not true. "Nothing" happens in the sense that the `map()` function will not raise any exception, it will simply iterate over the elements until it can't find a second argument to the function, at which point it simply stops and returns the result.

So, for example, if you evaluate `result = list(map(round, circle_areas, range(1, 3)))`, you won't get any error even as the length of `circle_areas` and the length of `range(1, 3)` differ. Instead, this is what Python does: It takes the first element of `circle_areas` and the first element of `range(1,3)` and passes it to `round`. `round` evaluates it then saves the result. Then it goes on to the second iteration, second element of `circle_areas` and second element of `range(1,3)`, round saves it again. Now, in the third iteration (`circle_areas` has a third element), Python takes the third element of `circle_areas` and then tries to take the third element of `range(1,3)` but since `range(1,3)` does not have a third element, Python simply stops and returns the result, which in this case would simply be `[3.6, 5.58]`.

Go ahead, try it.

In [ ]:
circle_areas = [3.56773, 5.57668, 4.00914, 56.24241, 9.01344, 32.00013]
result = list(map(round, circle_areas, range(1, 3)))

print(result)

The same thing happens if `circle_areas` is less than the length of the second iterable. Python simply stops when it can't find the next element in one of the iterables.

### `filter`

While `map()` passes each element in the iterable through a function and returns the result of all elements having passed through the function, `filter()`, first of all, requires the function to return boolean values (true or false) and then passes each element in the iterable through the function, "filtering" away those that are false. It has the following syntax:

`filter(func, iterable)`

The following points are to be noted regarding filter():

1. Unlike `map()`, only one iterable is required.
2. The `func` argument is required to return a boolean type. If it doesn't, `filter` simply returns the `iterable` passed to it. Also, as only one iterable is required, it's implicit that `func` must only take one argument.
3. `filter` passes each element in the iterable through `func` and returns only the ones that evaluate to true. I mean, it's right there in the name --- a "filter".

Let's see some examples

The following is a list (`iterable`) of the scores of 10 students in a Chemistry exam. Let's filter out those who passed with scores more than 75...using `filter`.

In [65]:
scores = [66, 90, 68, 59, 76, 60, 88, 74, 81, 65]

def is_A_student(score):
    return score > 75 # returns boolean (T or F)

over_75 = list(filter(is_A_student, scores))

print(list(filter(is_A_student, scores)))
print(over_75)

[90, 76, 88, 81]
[90, 76, 88, 81]


The next example will be a palindrome detector. A "palindrome" is a word, phrase, or sequence that reads the same backwards as forwards. Let's filter out words that are palindromes from a tuple (`iterable`) of suspected palindromes.

In [ ]:
dromes = ("demigod", "rewire", "madam", "freer", "anutforajaroftuna", "kiosk")

def check_palindrome(string):
  str_len = len(string)
  first, last = 0, str_len - 1
  while(first < last):
    if string[first] == string[last]:
      first = first + 1
      last = last - 1
    else:
      return False
  return True

palindromes = list(filter(check_palindrome, dromes))

print(palindromes)

### `reduce`

`reduce` applies a function of two arguments cumulatively to the elements of an iterable, optionally starting with an initial argument. It has the following syntax:

`reduce(func, iterable[, initial])`

Where `func` is the function on which each element in the `iterable` gets cumulatively applied to, and `initial` is the optional value that gets placed before the elements of the iterable in the calculation, and serves as a default when the iterable is empty. The following should be noted about `reduce()`: 1. `func` requires two arguments, the first of which is the first element in `iterable` (if `initial` is not supplied) and the second element in `iterable`. If `initial` is supplied, then it becomes the first argument to func and the first element in iterable becomes the second element. 2. `reduce` "reduces" (I know, forgive me) `iterable` into a single value.

As usual, let's see some examples.

Let's create our own version of Python's built-in `sum()` function. The `sum()` function returns the sum of all the items in the iterable passed to it.

In [ ]:
from functools import reduce

numbers = [3, 4, 6, 9, 34, 12]

def custom_sum(first, second):
    return first + second

result = reduce(custom_sum, numbers)
print(result)

So, what happened?

As usual, it's all about iterations: `reduce` takes the first and second elements in `numbers` and passes them to `custom_sum` respectively. `custom_sum` computes their sum and returns it to `reduce. reduce` then takes that result and applies it as the first element to `custom_sum` and takes the next element (third) in `numbers` as the second element to `custom_sum`. It does this continuously (cumulatively) until `numbers` is exhausted.

Let's see what happens when I use the optional `initial` value.

In [ ]:
from functools import reduce

numbers = [3, 4, 6, 9, 34, 12]

def custom_sum(first, second):
    return first + second

result = reduce(custom_sum, numbers, 10)
print(result)

## Higher Order Functions

What are higer-order functions?

A function that eiter:
* Takes another function as an argument
* Returns a function as its result

All other functions are considered first-order functions.

### Functions as Arguments

For example, let's consider the following computations:

\begin{align}
        &\sum_{k=1}^5 k   = 1 + 2 + 3 + 4 + 5 = 15 \\
        &\sum_{k=1}^5 k^2 = 1^2 + 2^2 + 3^2 + 4^2 + 5^2 = 55 \\
        &\sum_{k=1}^5 k^3 = 1^3 + 2^3 + 3^3 + 4^3 + 5^3 = 225
\end{align}

The common structure among functions may be a computational process, not just a number. A naive implementation of the above computational process would be:

In [ ]:
def sum_naturals(n):
  """Sum the first N natural numbers.

  >>> sum_naturals(5)
  15
  """
  total, k = 0, 1
  while k <= n:
    total, k = total + k, k + 1
  return total


def sum_squares(n):
  """Sum the first N squares of natural numbers.

  >>> sum_squares(5)
  55
  """
  total, k = 0, 1
  while k <= n:
    total, k = total + pow(k, 2), k + 1
  return total


def sum_cubes(n):
  """Sum the first N cubes of natural numbers.

  >>> sum_cubes(5)
  225
  """
  total, k = 0, 1
  while k <= n:
    total, k = total + pow(k, 3), k + 1
  return total

We can generalize the computational process as follows:

**함수를 인자로 받을 수 있음!! (term)**

In [ ]:
def identity(k):
  return k


def square(k):
  return k ** 2


def cube(k):
  return k ** 3


def summation(n, term):
  """sum the first N terms of a sequence.
  >>> summation(5, identity)
  15
  >>> summation(5, square)
  55
  >>> summation(5, cube)
  225
  """
  total = 0
  k = 1
  while k <= n:
    total = total + term(k)
    k = k + 1
  return total

In [66]:
def identity(k):
  return k

def square(k):
  return pow(k, 2)

def cube(k):
  return pow(k, 3)

def sigma(n, term):
  total = 0
  k = 1
  while(k<=n):
    total = total +term(k)
    k = k+1
  return total

print(sigma(5, identity))
print(sigma(5, square))
print(sigma(5, cube))

15
55
225


### Functions as Return Values

Functions defined within other function bodies are bound to names in a local frame:

In [ ]:
def make_adder(n):
  """Return a function that takes one argument k
     and returns k + n.
  >>> add_three = make_adder(3)
  >>> add_three(4)
  7
  """
  def adder(k):
    return k + n
  return adder

In [ ]:
print(type(make_adder(3)))
print(make_adder(3)(4))

func_add_three = make_adder(3)
print(func_add_three(7))

In [71]:
def make_adder(n):
  def adder(k): return k+n
  return adder

func_add_three = make_adder(3)
print(func_add_three(7))
print(make_adder(3)(7))

10
10


**WWPP Questions**

In [ ]:
def even(f):
  def odd(x):
    if x < 0:
      return f(-x)
    return f(x)
  return odd

def steven(x):
  return x

stewart = even(steven)
print(stewart)
print(stewart(61))
print(stewart(-4))

In [ ]:
def cake():
  print('beets')
  def pie():
    print('sweets')
    return 'cake'
  return pie

chocolate = cake()
print('1', chocolate)
print('2', chocolate())
more_chocolate, more_cake = chocolate(), cake
print('3', more_chocolate)

def snake(x, y):
  if cake == more_cake:
    return chocolate
  else:
    return x + y

print('4', snake(10, 20))
print('5', snake(10, 20)())
cake = 'cake'
snake(10, 20)

### Lambda Functions


A `lambda` function is a simple function definition that evaluates to a function:

`lambda <parameters>: <expression>`

For example,

In [ ]:
square = lambda x: x * x
# incorrect
# square = lambda x: return x * x

The difference between `def func` and `lambda` is
* `def` assigns a specific name to the function
* Both create a function (same domain, range and behavior)

We can simplifiy the examples we saw above as follows:

In [72]:
my_strings = ['a', 'b', 'c', 'd', 'e']
my_numbers = [1, 2, 3, 4, 5]

results = list(map(lambda x, y: (x, y), my_strings, my_numbers))

print(results)

[('a', 1), ('b', 2), ('c', 3), ('d', 4), ('e', 5)]


In [73]:
my_strings = ['a', 'b', 'c', 'd', 'e']
my_numbers = [1, 2, 3, 4, 5]

results = list(zip(my_strings, my_numbers))

print(results)

[('a', 1), ('b', 2), ('c', 3), ('d', 4), ('e', 5)]


In [74]:
summation(5, lambda k: k ** 1)
summation(5, lambda k: k ** 2)
summation(5, lambda k: k ** 3)

NameError: name 'summation' is not defined

Did you also notice that we didn't even need to create a function using the `def my_function()` standard way? That's how flexible Python in general is! Python allows to write less code (be "Pythonic").

**WWPP Questions**

In [ ]:
def add(a, b):
  return a + 5, b + 5

result = add(3, 2)
print(result)

In [ ]:
def outer_fun(a, b):
  def inner_fun(c, d):
      return c + d

  return inner_fun(a, b)
  return a

result = outer_fun(5, 10)
print(result)

In [ ]:
b = lambda x: lambda: x
c = b(88)
print(c)

In [ ]:
z = 3
e = lambda x: lambda y: lambda: x + y + z
print(e(0)(1)())

In [ ]:
higher_order_lambda = lambda f: lambda x: f(x)
g = lambda x: x * x

print(higher_order_lambda(2)(g))  # Which argument belongs to which function call?
print(higher_order_lambda(g)(2))

In [ ]:
call_thrice = lambda f: lambda x: f(f(f(x)))
call_thrice(lambda y: y + 1)(0)
print_lambda = lambda z: print(z)  # When is the return expression of a lambda expression executed?

print(print_lambda)
one_thousand = print_lambda(1000)
print(one_thousand)

## Style Guide for Python Code

Python style guides are a list of dos and don’ts for Python programs; other languages have their own style guides. Style guides let a team of programmers write codes that are **consistent**. The codes following a style guide seems just one programmer have wriiten the codes. For Python, two conventions are most popular
* [PEP-8](https://www.python.org/dev/peps/pep-0008/)
* [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html)